### PR TITLE
Add a hurry flag to the workflow spec.

### PR DESCRIPTION
**Pull request by roehrich-hpe**
The Slurm WLM has the concept of entering teardown state with the help of a "hurry" flag. This flag indicates that any in-progress data transfers must be terminated, or that any transfers not yet begun must be skipped.

It is left to the DWS driver implementation to determine how, and whether, to implement this flag.

Add constraints to the webhook to allow the hurry flag to be used only with teardown state.

Hook up the webhook in the controller tests. We need the reconciler running with the webhook to test things in teardown state.

Hook up tests in dockerfile so they run during the github workflow action, and enable tests in the workflow.

Signed-off-by: Dean Roehrich <dean.roehrich@hpe.com>

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -11,7 +11,7 @@ env:
   TEST_TARGET: testing
 
   # DO_TEST - true to build and run unit tests, false to skip the tests
-  DO_TEST: false
+  DO_TEST: true
```
Comment (ajfloeder) [inline .github/workflows/main.yml:14]: Thank you!

**Review by ajfloeder — APPROVED** [review]

File: api/v1alpha1/webhook_suite_test.go [hunk 2]
```diff
@@ -103,6 +103,7 @@ var _ = BeforeSuite(func() {
 	//+kubebuilder:scaffold:webhook
 
 	go func() {
+		defer GinkgoRecover()
```
Comment (bdevcich-hpe) [inline api/v1alpha1/webhook_suite_test.go:106]: Why is this needed?

File: api/v1alpha1/workflow_types.go [hunk 3]
```diff
@@ -137,7 +143,7 @@ type WorkflowDriverStatus struct {
 type WorkflowStatus struct {
 	// The state the resource is currently transitioning to.
 	// Updated by the controller once started.
-	// +kubebuilder:default=proposal
+	// +kubebuilder:default:=proposal
```
Comment (bdevcich-hpe) [inline api/v1alpha1/workflow_types.go:146]: In the old line, would this just not use the default then?

**Review by bdevcich-hpe — APPROVED** [review]

File: .github/workflows/main.yml (hunk 1 above)
Comment (NateRoiger) [reply .github/workflows/main.yml:14]: strange that this is even an option

Comment (billhpe) [reply .github/workflows/main.yml:14]: I implemented that when I first started working on github actions.  It was a more generic action and not everything had tests that existed or succeeded at that point in time.  It can be removed if no longer necessary.

**Review by matthew-richerson — APPROVED** [review]

File: api/v1alpha1/workflow_webhook.go [hunk 4]
```diff
@@ -76,8 +76,11 @@ var _ webhook.Validator = &Workflow{}
 // ValidateCreate implements webhook.Validator so a webhook will be registered for the type
 func (w *Workflow) ValidateCreate() error {
 
-	if w.Spec.DesiredState != "proposal" {
-		return fmt.Errorf("desired state must start in 'proposal'")
+	if w.Spec.DesiredState != StateProposal.String() {
+		return fmt.Errorf("desired state must start in %s", StateProposal.String())
+	}
+	if w.Spec.Hurry == true {
+		return fmt.Errorf("The hurry flag may not be set on creation")
```
Comment (NateRoiger) [inline api/v1alpha1/workflow_webhook.go:83]: Error strings should not be capitalized https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: controllers/suite_test.go [hunk 5]
```diff
@@ -72,13 +80,20 @@ var _ = BeforeSuite(func() {
...
+		Scheme:  scheme.Scheme,
+		Host:    webhookInstallOptions.LocalServingHost,
+		Port:    webhookInstallOptions.LocalServingPort,
+		CertDir: webhookInstallOptions.LocalServingCertDir,
 	})
 	Expect(err).NotTo(HaveOccurred())
 
 	// start reconcilers
 
+	err = (&dwsv1alpha1.Workflow{}).SetupWebhookWithManager(k8sManager)
```
Comment (NateRoiger) [inline controllers/suite_test.go:94]: Aren't all these tests in the controller redundant to what was done in the webhook test suite?

File: api/v1alpha1/webhook_suite_test.go (hunk 2 above)
Comment (roehrich-hpe) [reply api/v1alpha1/webhook_suite_test.go:106]: It allows ginkgo to catch failures from goroutines.  See https://onsi.github.io/ginkgo/#mental-model-how-ginkgo-handles-failure

I'm not sure that it matters much in this particular go routine.  I mainly just noticed that this is different from our other repos.

File: controllers/suite_test.go (hunk 5 above)
Comment (roehrich-hpe) [reply controllers/suite_test.go:94]: Two are redundant.  The first one I thought I'd leave in both.
I started putting them in the webhook test because that seemed logical, but then discovered I can't do much there without a status block, which doesn't happen if the reconciler isn't running so that's why I moved over there with the rest of the tests.

The second redundant test I'll remove from the webhook side...seems more convincing over on the controller side.

File: api/v1alpha1/workflow_types.go (hunk 3 above)
Comment (roehrich-hpe) [reply api/v1alpha1/workflow_types.go:146]: Looks like "make manifests" accepts either.  The docs show :=

**Review by billhpe — APPROVED** [review]
